File: Maximum Index - GFG/maximum-index.cpp
// { Driver Code Starts
#include <bits/stdc++.h>

using namespace std;

 // } Driver Code Ends
class Solution{
public:
    int maxIndexDiff(int arr[], int n) {
        // code here
        int res=INT_MIN;
        vector<int> left(n),right(n);
        long long mn=INT_MAX,mx=INT_MIN;
        for(int i=0;i<n;++i){
            if(arr[i]<mn)
                mn=arr[i];
            left[i]=mn;
        }
        for(int i=n-1;i>=0;--i){
            if(arr[i]>mx)
                mx=arr[i];
            right[i]=mx;
        }
        int i=0,j=0;
        while(i<n and j<n){
            if(left[i]<=right[j]){
                if(j-i>res)
                    res=j-i;
                j++;
            }
            else{
                i++;
            }
        }
        return res;
    }

};


// { Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, i;
        cin >> n;
        int a[n];
        for (i = 0; i < n; i++) {
            cin >> a[i];
        }
        Solution ob;
        auto ans = ob.maxIndexDiff(a, n);
        cout << ans << "\n";
    }
    return 0;
}  // } Driver Code Ends

File: Maximum Index - GFG/maximum-index.py
#User function Template for python3
class Solution:

	def maxIndexDiff(self,arr,n):
		#code here
		left=[0]*n
		right=[0]*n
		mn=float('inf')
		mx=float('-inf')
		for i in range(0,n):
		    mn=min(mn,arr[i])
		    left[i]=mn
	    for i in range(n-1,-1,-1):
		    mx=max(mx,arr[i])
		    right[i]=mx
		i=0
		j=0
		ans=float('-inf')
		while i<n and j<n:
		    if(left[i]<=right[j]):
		        ans=max(ans,j-i)
		        j+=1
            else:
                i+=1
        return ans
#{ 
#  Driver Code Starts
if __name__ == "__main__":
	t = int(input())
	while(t>0):
		num = int(input())
		arr = [int(x) for x in input().strip().split()]
		ob = Solution()
		print(ob.maxIndexDiff(arr,num))
		t-=1
# } Driver Code Ends